### PR TITLE
[PWA-1598] Add product to Cart from the wishlist

### PR DESCRIPTION
**Pull request by tjwiebell**
## Description

As a customer I want to add a wish-listed product to the cart *so that* I can buy a product that I saved for later earlier 
### Preconditions
 * Wishlist is enabled in Magento Admin Panel>Stores > Settings > Configuration > Customers > Wish List
 * Multi-wishlist is enabled in Stores > Settings > Configuration > Customers > Wish List>Enable Multiple Wish Lists (EE)
 * User is logged in to Venia and opens My Account>Favorites
 * User has products favorited

### Acceptance Criteria
 * User sees an option to add product to cart for each product in the wishlist. If configurable product is in the wishlist, user sees the option to configure it and add to the cart.
 * User selects adding product to the cart, cart gets updated and products stays in the favorite list. If user chooses to add the same product to the cart, product quantity gets updated.
 * If product in the favorites is out of stock, user sees an indication of that, no option to add it wishlist available.

### Test Plan
1. Add a Simple item to wishlist
2. Verify that you can add the simple item to the cart successfully without having to select any options
3. Verify that the item also remains in the wishlist
4. Add a configurable item with options selected to wishlist
5. Verify that you can add the configurable item to the cart successfully (have to confirm option selections?)
6. Verify that the item also remains in the wishlist
7. Add a configurable item with no options selected to wishlist
8. Verify that you can add the item to the cart successfully, having to first select item options via a dialog
9. Verify that the dialog works on mobile
10. Verify that the item remains in the wishlist after being added to cart
11. Mark one of the wishlist items as out of stock in the Magento Admin
12. Verify that the item's "Add to Cart" button is disabled. It should be visually different (grayed out), and not clickable

## Related Issue

* [[PWA-1598](https://jira.corp.magento.com/browse/PWA-1598)] Add product to Cart from the wishlist

### Verification Steps

1. See Test Plan

## Checklist

* I have added tests to cover my changes, if necessary.
* I have added translations for new strings, if necessary.
* I have updated the documentation accordingly, if necessary.

### CODE REVIEW

File: packages/venia-ui/lib/components/WishlistPage/wishlistItem.js [hunk 1]
```diff
@@ -67,65 +59,66 @@ const WishlistItem = props => {
...
 
     return (
         <div className={classes.root}>
-            <Image
-                alt={imageLabel}
-                classes={{ image: classes.image }}
-                src={url}
-                width={400}
-            />
+            <Image {...imageProps} />
```
Comment (sirugh) [inline packages/venia-ui/lib/components/WishlistPage/wishlistItem.js:92]: For some reason the first image in my list is lower resolution and blurry. I can give you the account credentials to check against. The same account on `develop` does not have this issue.

[Image from Gyazo](https://gyazo.com/4aeb771e1244c8191fa953a6ab9a95a6)

**Review by sirugh — CHANGES_REQUESTED** [review]
Good work - not much to comment on here, though I did notice something off with the images (or at least one product's images). You'll also want to remedy the query validator failure. Also the view-specific talon props are an interesting pattern and I'd be interested to hear the team's take on it.

File: packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        }
+
+        return [];
+    }, [item, userSelectedOptions]);
+
+    const { data } = useQuery(operations.getProductDetailQuery, {
+        fetchPolicy: 'cache-and-network',
+        nextFetchPolicy: 'cache-first',
+        variables: {
+            configurableOptionValues: selectedOptionsArray,
```
[the remark this answers is not in the document]
Comment (tjwiebell) [reply packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js:58]: That is a little surprising, I thought Apollo did its own memoization and identity itself shouldn't be a trigger. Possible the order of the value_uid's changed, but I'll investigate just out of curiosity.

File: packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                    ? selectedProductMediaGallery[0]
+                    : product.image;
+
+            setCurrentImage(currentImage);
+
+            const finalPrice = selectedVariant
+                ? selectedVariant.price_range.maximum_price.final_price
+                : product.price_range.maximum_price.final_price;
+
+            setCurrentPrice(finalPrice);
```
[the remark this answers is not in the document]
Comment (tjwiebell) [reply packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js:88]: I tried to pull this out of PO/UX and they were comfortable with the delayed re-render with no indicator. I would love to have something here if you have any simple ideas. Maybe something like the page-level loading indicator, but in the header of the Dialog?

File: packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        if (item) {
+            return {
+                disabled:
+                    item.product.configurable_options.length !==
+                        selectedOptionsArray.length || isAddingToCart,
+                onClick: handleAddToCart,
+                priority: 'high'
+            };
+        }
+    }, [handleAddToCart, isAddingToCart, item, selectedOptionsArray.length]);
```
[the remark this answers is not in the document]
Comment (tjwiebell) [reply packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js:163]: Discussed in Office Hour, going to run with it for now as it should give community easier methods of extending component behavior via talon interception.

File: packages/venia-ui/lib/components/WishlistPage/wishlistItem.js (hunk 1 above)
Comment (tjwiebell) [reply packages/venia-ui/lib/components/WishlistPage/wishlistItem.js:92]: I've seen this once before, and it was from switching between mobile and desktop. If the browser caches a lower res image when in mobile, it'll re-use it when you switch to desktop. Doesn't seem like a realistic use case, but I'll see what the options are to remedy.

File: packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        variables: {
+            configurableOptionValues: selectedOptionsArray,
+            sku
+        },
+        skip: !sku
+    });
+
+    const [addProductToCart, { loading: isAddingToCart }] = useMutation(
+        operations.addProductToCartMutation
+    );
```
[the remark this answers is not in the document]
Comment (tjwiebell) [reply packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js:66]: Yes, yes we do. Good callout 👍

File: packages/venia-ui/lib/components/WishlistPage/wishlistItem.js (hunk 1 above)
Comment (sirugh) [reply packages/venia-ui/lib/components/WishlistPage/wishlistItem.js:92]: Yea I didn't know how to repro so if it doesn't occur out of "normal" use I wouldn't worry about it.

File: packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (sirugh) [reply packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js:88]: > Maybe something like the page-level loading indicator, but in the header of the Dialog?

This is a good idea! I think a "shimmering" gradient in the background of the dialog header would be cool.

[the remark this answers is not in the document]
Comment (tjwiebell) [reply packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js:92]: Went simple and added the loading indicator to header in 92e8ffd.

File: packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (tjwiebell) [reply packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js]: Added in 92e8ffd.

File: packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sirugh) [reply packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js]: Did you solve this? I see you added `id` to variant and configurable option.

File: packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (sirugh) [reply packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js:92]: I saw - does it bother you that it is shifted to the right and spins slightly off center?

[Image from Gyazo](https://gyazo.com/886fde543ad44ef0d76b36040507fe97)

File: packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (tjwiebell) [reply packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js]: No, couldn't figure this out. It falls back to doing `cache-and-network` like it's the first fetch, even though the options array is equivalent (except identity). I could retain the identity of the array and mutate, but I'm not sure if that would trigger fetches when the options do change. I'm curious and will continue digging, but wouldn't block on it.

**Review by sirugh — APPROVED** [review]

**Review by sirugh — DISMISSED** [review]
Thanks :)

File: packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (tjwiebell) [reply packages/peregrine/lib/talons/AddToCartDialog/useAddToCartDialog.js:92]: Most definitely, fixed in 7f53219.

**Comment by dpatil-magento** [comment]
QA Approved.

**Review by sirugh — APPROVED** [review]

**Pull request merged**